presto-main/src/main/java/io/prestosql/sql/planner/optimizations/QueryCardinalityUtil.java

```diff
@@ -106,10 +106,33 @@ public CardinalityExtractorPlanVisitor(Lookup lookup)
         @Override
         public Range<Long> visitAggregation(AggregationNode node, Void context)
         {
-            if (node.hasEmptyGroupingSet()) {
+            if (node.hasEmptyGroupingSet() && node.getGroupingSetCount() == 1) {
+                // only single default aggregation which will produce exactly single row
                 return Range.singleton(1L);
             }
-            return Range.atLeast(0L);
+
+            Range<Long> sourceCardinalityRange = node.getSource().accept(this, null);
+
+            if (node.hasDefaultOutput() || sourceCardinalityRange.lowerEndpoint() > 0) {
```

Comment:
Why do we need `hasDefaultOutput()`? Checking for `hasEmptyGroupingSet` should be sufficient since that's the only case that will guarantee at least one row without input, no? (In fact, I'm not sure we need hasDefaultOutput at all in AggregationNode anymore, but that's for a separate discussion)

Reply:
> Why do we need hasDefaultOutput()? 

Only partial and single aggregations produce default output. Final aggregations do not. That's what this check is about (even though it's not super elegant as it depends how execution works).